stock_available_to_promise_release/models/stock_move.py

```diff
@@ -427,3 +429,54 @@ def _assign_picking_post_process(self, new=False):
         if priorities:
             self.picking_id.write({"priority": max(priorities)})
         return res
+
+    def _get_all_orig_ids(self):
+        """Return a dictionary by move id of the ids of all the moves that are
+        the origin of the current recordset
+        """
+        return self._get_all_moves_from_chain("move_orig_ids")
+
+    def _get_all_dest_ids(self):
+        """Return a dictionary by move id of the ids of all the moves that are
+        the destination of the current recordset
+        """
+        return self._get_all_moves_from_chain("move_dest_ids")
+
+    def _get_all_moves_from_chain(self, field):
+        """Return a dictionary by move id of all the moves that are the origin
+        or destination of this move
+        """
+        res = dict.fromkeys(self.ids, [])
+        if not self:
+            return {}
+        select_col = field[:-1]
+        link_col = "move_orig_id" if field == "move_dest_ids" else "move_dest_id"
+        SQL = """
+            WITH RECURSIVE all_moves AS (
+                SELECT
+                    %(select_col)s AS id,
+                    %(link_col)s AS move_id
+                FROM
+                    stock_move_move_rel
+                WHERE
+                    %(link_col)s in %(move_ids)s
+                UNION ALL
+                    SELECT
+                        %(select_col)s AS id,
+                        %(link_col)s AS move_id
+                    FROM
+                        stock_move_move_rel
+                    JOIN all_moves ON all_moves.id = stock_move_move_rel.%(link_col)s
+            )
+            SELECT move_id, ARRAY_AGG(id) FROM all_moves GROUP BY move_id
+        """
+        self.env.cr.execute(
+            SQL,
+            {
+                "select_col": AsIs(select_col),
+                "link_col": AsIs(link_col),
+                "move_ids": tuple(self.ids),
+            },
+        )
+        res.update(dict(self.env.cr.fetchall()))
```

Comment:
The chain of moves has just been created. Aren't all ids already in the cache? Do you have less queries with this change?

Reply:
The chained moves are not into the cache in case of unrelease and when computing if an unreleased is allowed. The goal is to be sure that with just 1 additional SQL query we've all the information required to compute in batch the `unrelease_allowed`field regardless of the moves chain depth. The benefit may not be enough compared to the complexity. Furthermore, it is true that such a query requires an explicit flush of the cache for the stock.move model and so, not sure if this is a real benefit in the end. I'll change the code.